plutus-core/plutus-core/src/PlutusCore/Builtin/Convert.hs

```diff
@@ -0,0 +1,526 @@
+>         subroutineLE arg3 = ...
+>         subroutineBE arg3 = ...
+
+This form _ensures_ we scrutinize (and branch) only the number of times we have
+to, and in a predictable place. Since these are backends for Plutus Core primops,
+and performance is thus critical, we choose to use this manually-specialized form
+for each combination of relevant arguments. While this is repetitive, and thus
+also somewhat error-prone, the performance penalty for not doing this is
+unacceptable.
+-}
+
+{- Note [Loop sectioning]
+Both integerToByteString and byteStringToInteger effectively function as loops
+over digits (and thus, individual bytes), which either have to be read or
+extracted. In particular, this involves trafficking data between memory and
+machine registers (both ByteString and Integer are wrappers around counted
+arrays), as well as the overhead of looping (involving comparisons and branches).
+
+However, on all architectures of interest (essentially, 64-bit Tier 1),
+general-purpose registers (GPRs henceforth) are 64 bits (or 8 bytes).
+Furthermore, the primary cost of moving data between memory and registers is
+having to overcome the 'memory wall': the exact amount of data being moved
+doesn't affect this very much. In addition to this, when we operate on single
+bytes, the remaining 56 bits of the GPR holding that data are essentially
+'wasted'. In the situation we have (namely, operating over arrays, whose data
+is adjacent in memory), we thus get two sources of inefficiency:
+
+* Despite paying the cost for a memory transfer, we transfer only one-eighth
+  the data we could; and
+* Despite transferring data from memory to registers, we utilize the register
+  at only one-eighth capacity.
+
+This essentially means we perform eight times _more_ rotations of the loop
+than we need to!
+
+To avoid this inefficiency, we use a technique known as _loop sectioning_.
+Effectively, this turns our homogenous loop (that always operates one byte at
+a time) into a heterogenous loop: first, we operate on a larger section (called
+a _stride_) until we can no longer do this, and then we finish up using byte
+at a time processing. Essentially, when given an input like this:
+
+[ b1, b2, b3, b4, b5, b6, b7, b8, b9, b10 ]
+
+the homogeous byte-at-a-time approach would process it like so:
+
+  _   _   _   _   _   _   _   _   _   _
+[ b1, b2, b3, b4, b5, b6, b7, b8, b9, b10 ]
+
+for a total of 10 memory transfers and 10 loop spins, whereas a loop-sectioned
+approach with a stride of 8 would instead process like so:
+
+  ______________________________  _   _
+[ b1, b2, b3, b4, b5, b6, b7, b8, b9, b10 ]
+
+Giving us only _three_ memory transfers and _three_ loop spins instead. This
+effectively reduces our work by a factor of 8. In our cases, this is almost
+free, as there is no data processing to be done: all we need to do is copy
+from one place to another, essentially.
+
+This technique only benefits us because counted arrays are cache-friendly: see
```

Comment:
Yes, this _definitely_ matters a lot. My original version of both functions didn't loop section at all, and I actually got _more_ than a factor-of-8 speedup when I did the sectioning, as I reduced the number of integer operations by a factor of 8 (and they're linear), as well as the number of copy operations that'd be required (and they're _also_ linear).

This is especially important, as I can't (easily) access the representation of `Integer`s directly while being compatible with GHC 8.10. If and when 8.10 gets jettisoned, we can avoid all of this and just directly copy memory.

Reply:
Since it sounds like you've already written the fast implementation, should we include it behind some CPP so it's ready to go later?

Reply:
We discussed this with @kwxm, and decided against it, as it would mean that different nodes would run the same program with quite different costings. Due to the approach I'm taking here (to be 8.10 compatible), the costing is quadratic (`Integer` operation linearity forces our hand), but a direct copying method is linear.

Unless you mean just including it in the codebase behind an effective 'do not compile this'-CPP?

Reply:
Sorry yes, I meant the latter. It can be some dead code that we can use later. It needs the CPP only because it presumably won't compile on 8.10.

Reply:
It won't compile as-is on 9.2 either. Furthermore, the CIP has changed in a significant-enough way since that the implementation would need rethinking anyway. I don't think there's much reason to include my prior work as-currently-is: might be worth coming back to this question once 8.10 is gone.